[FLINK-12201][network,metrics] Introduce InputGateWithMetrics in Task to increment numBytesIn metric

zhijiangW opened the pull request:
## What is the purpose of the change

* Incrementing of `numBytesIn` metric in `SingleInputGate` does not depend on shuffle service and can be moved out of network internals into `Task`. `Task` could wrap `InputGate` provided by `ShuffleService` with `InputGateWithMetrics` which would increment numBytesIn metric.*

## Brief change log

  - *Introduce `InputGateWithMetrics` to wrap `InputGate` from `ShuffleService` in `Task`*
  - *Remove the counter from `SingleInputGate` and related creation.*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented:
@flinkbot attention @azagrebin

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -191,7 +191,7 @@
 	private final ResultPartition[] producedPartitions;
 
-	private final SingleInputGate[] inputGates;
+	private final InputGate[] inputGates;
```
I would keep it as `SingleInputGate[]` to avoid that copying loop `SingleInputGate[]` -> `InputGate[]` in Task constructor at the moment.

Later, we could move `inputGatesById` completely to `NetworkEnviroment`, when we refactor:
- `TaskExecutor.updatePartitions` to use `NetworkEnviroment.update(PartitionInfo)`
- try to move `triggerPartitionProducerStateCheck` to `SingleInputGate` and pass `partitionProducerStateChecker` to `SingleInputGate` instead of having it in `TaskActions`

`inputGatesById` looks like the last blocker to change `SingleInputGate[]` -> `InputGate[]` in Task.

zhijiangW replied:
I agree with the following refactors for `updatePartitions` and `triggerPartitionProducerStateCheck` which could help remove `inputGatesById` in task.

But the change of `SingleInputGate[]` to `InputGate[]` is necessary here because we have to loop the `SingleInputGate[]` from `NetworkEnvironment#createInputGates` to generate `InputGateWithMetrics` which implements `InputGate`. The task would finally maintain `InputGate[]` instead of `SingleInputGate`, so this change would be consistent with future plans.

Did I misunderstand your concern?

azagrebin replied:
Oh, sorry, you are right, too fast review :)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferOrEvent.java:
```diff
@@ -43,26 +46,30 @@
 		this.event = null;
 		this.channelIndex = channelIndex;
 		this.moreAvailable = moreAvailable;
+		this.size = size;
```
Yes, I think it is available before.

azagrebin commented:
@flinkbot approve all

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java:
```diff
@@ -0,0 +1,119 @@
+	@Override
+	public Optional<BufferOrEvent> getNextBufferOrEvent() throws IOException, InterruptedException {
+		Optional<BufferOrEvent> bufferOrEvent = inputGate.getNextBufferOrEvent();
+		if (bufferOrEvent.isPresent()) {
```
As an alternative:
`bufferOrEvent.ifPresent(b -> numBytesInCounter.inc(b.getSize()));`

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	@Override
+	public Optional<BufferOrEvent> pollNextBufferOrEvent() throws IOException, InterruptedException {
```
It seems we also need to increment here.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java:
```diff
@@ -0,0 +1,119 @@
+ * shuffle service. It is mainly used for increasing general input metrics from
+ * {@link TaskIOMetricGroup}.
+ */
+public class InputGateWithMetrics implements InputGate, InputGateListener {
```
Not sure why we need to implement `InputGateListener` as well.
Could we not just forward `registerListener` to `inputGate`?
Who is going to call `notifyInputGateNonEmpty` in this class?

zhijiangW replied:
I also not considered this issue before, only found it when existing ITCase fails.

The `UnionInputGate` could only see `InputGateWithMetrics` and then put them into internal `inputGateToIndexOffsetMap`. When the `SingleInputGate` receives data from network, it would notify the listener(`UnionInputGate`) and then the `SingleInputGate` would be added into `UnionInputGate#inputGatesWithData`. When `UnionInputGate#getNextBufferOrEvent`, the `SingleInputGate` is fetched from `inputGatesWithData`, but it can not be found from `inputGateToIndexOffsetMap` which would cause NPE issue.

azagrebin replied:
I see, thanks for explanation!

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	@Override
+	public void close() throws Exception {
```
should we not forward `close` to `inputGate` as well?

zhijiangW replied:
In task class we would finally maintain only the array of `InputGateWithMetrics` finally, so the close called from task would focus on `InputGateWithMetrics`. Otherwise the task would maintain two arrays `SingleInputGate` and `InputGateWithMetrics` separately.

zhijiangW replied:
If we change to wrap the `InputGateWithMetrics` in the constructor as you mentioned below, we should forward `close` to `InputGate` as well.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java:
```diff
@@ -0,0 +1,119 @@
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The InputGateWithMetrics is wrapped in task with {@link InputGate} provided by
```
`This class wraps {@link InputGate} provided by ...`

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -637,6 +636,11 @@ else if (current == ExecutionState.CANCELING) {
 			TaskKvStateRegistry kvStateRegistry = kvStateService.createKvStateTaskRegistry(jobId, getJobVertexId());
 
+			InputGate[] gates = new InputGate[inputGates.length];
+			for (int i = 0; i < inputGates.length; i++) {
```
To be honest, I liked wrapping in constructor more. This way we keep in `Task` and `RuntimeEnvironment` same objects and there is less risk for them to diverge and forget that we call real gate and not wrapper. What do you think?

I would at least move the wrapping loop to a separate static method.

zhijiangW replied:
Yes, I also did it in the constructor in first version. Later I think it might be better to make `InputGateWithMetrics` involved in less logics such as `close`, so I change it as now to keep the previous arrays of `SingleInputGate` in task.

I think it also makes sense to do it in constructor, then I would fix it.

zhijiangW replied:
Forgot another concern for not doing in the constructor. It would make maintain two arrays of `SingleInputGate` and `InputGate` separately. Because we still need to return the array of `SingleInputGate` for `NetworkEnvironment#setup` currently. What do you think?

zhijiangW commented:
@azagrebin I have rebased the master to solve the conflicts.

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -261,8 +260,7 @@ private InputGate createInputGate(
 					jobId,
 					gateDescriptor,
 					new NoOpTaskActions(),
-					InputChannelTestUtils.newUnregisteredInputChannelMetrics(),
```
We should actually also wrap `gate` here to take wrapping into account in benchmark tests. Probably with a dummy metric counter, at least forcing the wrapper to be called.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -382,19 +382,21 @@ public Task(
 
 		// consumed intermediate result partitions
-		this.inputGates = networkEnvironment.createInputGates(
+		SingleInputGate[] gates = networkEnvironment.createInputGates(
```
Yes, no matter which one is merged first, then the other one has to rebase to solve the conflicts for concerning this issue. :)

zhijiangW commented:
Thanks for further reviews @azagrebin !

I pushed one fixup commit for addressing above comments!  And I would submit the micro benchmark comparison later.

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -249,18 +250,21 @@ private InputGate createInputGate(
+				environment.getNetworkBufferPool());
+
+			final InputGate gate = new InputGateWithMetrics(
+				gateFactory.create(
```
nit: maybe have a separate variable for `gateFactory.create`
also channel loop body could be refactored into two methods: `createInputGateDeploymentDescriptor` and `createSingleInputGateWithMetrics` to reduce the current method

zhijiangW replied:
Actually gate factory should not be created inside loop, and it could be created only once together with `NetworkEnvironment`. I would submit a hotfix commit firstly for refactoring this method, and then append my PR based on it.

zhijiangW commented:
I have squashed the commits for addressing the comments. @azagrebin

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -227,39 +229,22 @@ protected ResultPartitionWriter createResultPartition(
-			ExecutionAttemptID executionAttemptID,
 			final TaskManagerLocation senderLocation,
-			NetworkEnvironment environment,
+			final SingleInputGateFactory gateFactory,
```
is `gateFactory` not already available as a class field in the method?
strange that the field is protected, can be private

zhijiangW replied:
I also found this issue before, but considering all the fields in this class are defined as protected, so I keep the way as others. 
Also for the class field factory, I keep the way as field `channels`  to pass because it is not initialized in the constructor. Actually I also think many other fields in this class could not be passed as well.
Do you think we should further refactoring this?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java:
```diff
@@ -0,0 +1,106 @@
+
+	@Override
+	public Optional<BufferOrEvent> getNextBufferOrEvent() throws IOException, InterruptedException {
+		Optional<BufferOrEvent> bufferOrEvent = inputGate.getNextBufferOrEvent();
```
```
getNextBufferOrEvent() {
  return updateMetrics(inputGate.getNextBufferOrEvent());
}

pollNextBufferOrEvent() {
  return updateMetrics(inputGate.pollNextBufferOrEvent());
}
```

to deduplicate the code a bit and decrease chances of methods being out of sync (as pointed by @azagrebin in another comment)?

zhijiangW replied:
We could deduplicate one common line here, but I am not clear  why decrease changes out of sync. Are there any syncs related to here?

pnowojski replied:
Not that I'm aware of :) I meant for the future, to avoid similar bugs/issues as pointed out by @azagrebin

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferOrEvent.java:
```diff
@@ -43,26 +44,32 @@
-	BufferOrEvent(Buffer buffer, int channelIndex, boolean moreAvailable) {
+	private final int size;
+
+	public BufferOrEvent(Buffer buffer, int channelIndex, boolean moreAvailable, int size) {
```
why are we passing this new field explicitly? Why not
```
this.size = buffer.getSize();
```
?

Further more, why do we need `size` field at all? Why not implement:
```
int getSize() {
  return buffer == null ? 0 : buffer.getSize();
}
```

zhijiangW replied:
For event buffer, we might not return 0, so we have the explicit size field, but we do not need pass the size for buffer constructor as you suggested.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -566,9 +560,9 @@ private BufferOrEvent transformToBufferOrEvent(
 			boolean moreAvailable,
 			InputChannel currentChannel) throws IOException, InterruptedException {
-		numBytesIn.inc(buffer.getSizeUnsafe());
+		final int size = buffer.getSizeUnsafe();
```
Hmm, it looks like we can drop `getSizeUnsafe()` method altogether, since it doesn't differ from `getSize()` - both are not synchronized (also like `setSize()`). It seems like I or @NicoK missed that when we were implementing low latency changes and dropping the synchronisation in the `Buffer` class. Or am I missing something?

zhijiangW replied:
Yes, it could be replaced by `getSize` here. I could submit a separate hotfix commit for it?

pnowojski replied:
Yes, separate hotfix commit would be best (sorry for nit picking here, just noticed this here, since I got curious whether you should be using unsafe or safe version)

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -270,4 +266,18 @@ private InputGateDeploymentDescriptor createInputGateDeploymentDescriptor(
+			SingleInputGateBuilder.NO_OP_PRODUCER_CHECKER,
+			InputChannelTestUtils.newUnregisteredInputChannelMetrics());
+
+		return new InputGateWithMetrics(singleGate, new SimpleCounter());
```
Could we somehow deduplicate the code of wrapping the input gate between production and benchmarking code? So that they follow the same path? For example moving the wrapping for both code paths to the `SingleInputGateFactory`?

zhijiangW replied:
Yes, we could make `SingleInputGateFactory` always return wrapped `InputGateWithMetrics`

zhijiangW replied:
There has one concern here. 
`SingleInputGateFactory` is `NetworkEnvironment`/`ShuffleService` specific, not a general factory. `InputGateWithMetrics` is not belonging to `ShuffleService`, and it should be in task scope. So it might be not proper to let `SingleInputGateFactory` wrap for task related components, otherwise it seems no need to introduce `InputGateWithMetrics` in this PR. WDYT?

azagrebin replied:
I agree that `InputGateWithMetrics` does not belong Netty implementation of shuffle service this is why we are doing it change to keep this notification concern outside.

pnowojski replied:
Ok, I understand. In that case maybe we could consider some another layer of factory/builder? One that takes a concrete NetworkEnvironment and creates InputGates wrapped with metrics? I'm not sure if that's worth, but you see the code duplication here? And the risks that if those two code parts will drift apart, the benchmarks might become inaccurate.

zhijiangW replied:
I could understand your concern here. 
ATM the `InputGateWithMetrics` is only created in the benchmark which would reflect the real process. Considering this wrapper is very easy to construct and the duplication here is not obvious, just a local `singleGate` definition, I think we could introduce another factory for this wrapped `InputGate` if necessary later?

pnowojski replied:
Ok

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:
```diff
@@ -158,17 +158,6 @@
 	 */
 	void setReaderIndex(int readerIndex) throws IndexOutOfBoundsException;
 
-	/**
```
Nit: typo in the commit message (reductant -> redundant)
